### PR TITLE
Postinst: do not ask for password if it was already specified

### LINKED ISSUE #8316: Conflicting user configuration files created during update
After updating Clickhouse as described in https://clickhouse.yandex/docs/en/operations/update/ from version 19.3.5 to 19.17.4.11, clickhouse-server service refused to start. I have Ubuntu 18.04.1 LTS.

**How to reproduce**
* Clickhouse version - 19.3.5, user settings stored in `/etc/clickhouse-server/users.xml`. Password for user "default" is stored in `<password_sha256_hex></password_sha256_hex>` block
* Updating to the newest version (19.17.4.11 in my case)
* During the update system asks for password for default user; I entered it
* Choose to keep previously created config files
* Restarting clickhouse-server daemon -> service doesn't start

**Expected behavior**
Service "clickhouse-server" restarts without errors after update.

**Error message and/or stacktrace**
When trying to start clickhouse-server:
`Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = Cannot create file: Cannot create pid file. (version 19.17.5.18 (official build)`

In `/var/log/clickhouse-server/clickhouse-server.err.log`: 
`<Error> Application: DB::Exception: More than one field of 'password', 'password_sha256_hex', 'password_double_sha1_hex' is used to specify password for user default. Must be only one of them.`

**How I solved the problem**
Deleted "users.d" folder with conflicting settings that was created after an update.

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not ask for password for default user during installation of debian package if the password was already specified in condiguration but this fact was not saved earlier inside debian tools. This closes #8316.

### CODE REVIEW

File: debian/clickhouse-server.postinst [hunk 1]
```diff
@@ -128,18 +129,22 @@ Please fix this and reinstall this package." >&2
...
-    if [ -f /usr/share/debconf/confmodule ]; then
-        db_get clickhouse-server/default-password
-        defaultpassword="$RET"
-        if [ -n "$defaultpassword" ]; then
-            echo "<yandex><users><default><password>$defaultpassword</password></default></users></yandex>" > ${CLICKHOUSE_CONFDIR}/users.d/default-password.xml
-            chown ${CLICKHOUSE_USER}:${CLICKHOUSE_GROUP} ${CLICKHOUSE_CONFDIR}/users.d/default-password.xml
-            chmod 600 ${CLICKHOUSE_CONFDIR}/users.d/default-password.xml
-        fi
+    # Ask for password for default user. But only if it was not specified in configuration.
+    if [ -z "$($(su -s $SHELL ${CLICKHOUSE_USER} -c "$CLICKHOUSE_BINDIR/$EXTRACT_FROM_CONFIG --config-file=\"$CLICKHOUSE_USERS\" --key=users.default.password" 2>/dev/null; su -s $SHELL ${CLICKHOUSE_USER} -c "$CLICKHOUSE_BINDIR/$EXTRACT_FROM_CONFIG --config-file=\"$CLICKHOUSE_USERS\" --key=users.default.password_sha256_hex" 2>/dev/null) ||: )" ]; then
```
Comment (alesapin) [inline debian/clickhouse-server.postinst:133]: Maybe move this quite long line out of if clause and store both values in a separate variable?

Comment (alexey-milovidov) [reply debian/clickhouse-server.postinst]: Ok.

The main concern is that I cannot add a test and have to test it manually just once before merging...

Comment (alexey-milovidov) [reply debian/clickhouse-server.postinst]: Also Debian already remembers that the password was specified.
And the fix is only for corner case when the password already specified in config but clickhouse-server package was not installed before.

**Comment by alexey-milovidov** [comment]
Unfortunately CI did not create a build for this PR.

**Comment by alexey-milovidov** [comment]
I will modify some unrelated file for this purpose...

**Comment by alexey-milovidov** [comment]
> Yandex synchronization check (only for Yandex employees) Pending — error, check will be restarted

This check is unreliable @exprmntr

**Comment by alexey-milovidov** [comment]
This modification is practially untestable. The only way is to test manually... I have tested and it does not work.

**Comment by alexey-milovidov** [comment]
Test 1:

Install debian package and check that it's asking for password.
It should not break with this PR.

Test 2:

Install ClickHouse without debian package (by any other means), set up default password either with plaintext or with sha256 (both options should be tested). Then install ClickHouse with debian package and check that it does not ask for password.

Test 3:

Install ClickHouse with debian package, do dpkg purge but save config file with password. Then install again and check that it does not ask for password.

**Comment by alexey-milovidov** [comment]
I've changed "bugfix" to "improvement" so no one will dare to backport this insignificant mostly unneeded useless fix.

**Comment by akuzm** [comment]
> I've changed "bugfix" to "improvement" so no one will dare to backport this insignificant mostly unneeded useless fix.

The bots fight you again... better use `pr-no-backport`.

**Comment by alexey-milovidov** [comment]
debconf does not follow script execution. Instead it parses the script and tries to ask for configuration for all options listed inside:
```
root@4b0efbc31f64:/packages# ar x clickhouse-server_20.8.1.4478_all.deb 
root@4b0efbc31f64:/packages# bash -x ./postinst 
+ set -e
+ CLICKHOUSE_USER=clickhouse
+ CLICKHOUSE_GROUP=clickhouse
+ CLICKHOUSE_CONFDIR=/etc/clickhouse-server
+ CLICKHOUSE_DATADIR=/var/lib/clickhouse
+ CLICKHOUSE_LOGDIR=/var/log/clickhouse-server
+ CLICKHOUSE_BINDIR=/usr/bin
+ CLICKHOUSE_GENERIC_PROGRAM=clickhouse
+ EXTRACT_FROM_CONFIG=clickhouse-extract-from-config
+ CLICKHOUSE_CONFIG=/etc/clickhouse-server/config.xml
+ CLICKHOUSE_USERS=/etc/clickhouse-server/users.xml
+ '[' -f /usr/share/debconf/confmodule ']'
+ . /usr/share/debconf/confmodule
++ '[' '!' '' ']'
++ PERL_DL_NONLAZY=1
++ export PERL_DL_NONLAZY
++ '[' '' ']'
++ exec /usr/share/debconf/frontend ./postinst
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (Can't locate Term/ReadLine.pm in @INC (you may need to install the Term::ReadLine module) (@INC contains: /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.30.0 /usr/local/share/perl/5.30.0 /usr/lib/x86_64-linux-gnu/perl5/5.30 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.30 /usr/share/perl/5.30 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at /usr/share/perl5/Debconf/FrontEnd/Readline.pm line 7.)
debconf: falling back to frontend: Teletype
Enter password for default user:
```

**Comment by alexey-milovidov** [comment]
> Note: These questions are asked by a separate config script, not by the postinst, so the package can be configured before it is installed, or reconfigured after it is installed. Do not make your postinst use debconf to ask questions.

http://www.fifi.org/doc/debconf-doc/tutorial.html

**Pull request closed without merging**